### PR TITLE
[Inference API] OpenAI Completions API

### PR DESCRIPTION
**Pull request by timgrein**
This PR adds support for a new task type named `COMPLETION` with OpenAI as the first implemented provider for completions.
The integrated API is [OpenAI's chat completions API](https://platform.openai.com/docs/guides/text-generation/chat-completions-api). Currently only one prompt with one response is supported as we should think about what multiple inputs imply (a conversation/context? multiple single prompts independent from each other?) before supporting them IMO.

Sample requests:

**Put OpenAI completion model**:

```
PUT {{ES_HOST}}/_inference/completion/openai_chat_completions

{
    "service": "openai",
    "service_settings": {
        "api_key": "{{OPEN_AI_API_KEY}}",
        "model_id": "gpt-3.5-turbo"
    }
}
```

**Perform Summarization**:

```
POST {{ES_HOST}}/_inference/completion/openai_chat_completions

{
  "input": "Summarize the following text: text..."
}
```

**Perform Q&A**:
```
POST {{ES_HOST}}/_inference/completion/openai_chat_completions

{
  "input": "Answer the following question: question"
}
```

**400 error for multiple inputs**:
```
POST {{ES_HOST}}/_inference/completion/openai_chat_completions

{
    "input": [
        "What is Elastic?",
        "What is MongoDB?"
    ]
}
```

### CODE REVIEW

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiService.java
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiService.java:195]: How about we move this check to `OpenAiChatCompletionAction`. That way we can avoid the `instanceof` check.

**Review by jonathan-buttner — COMMENTED** [review]
Great work! I left a few questions and comments.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java:145]: Also when you add the transport version, let's bump the version here: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiService.java#L257

That way when a cluster is in the process of upgrading, we won't let a user create a new completion openai service until the cluster is fully upgraded. This avoids the user running into an issue if they create a new openai service and try to use the completion functionality before the other nodes are aware of it.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiChatCompletionResponseEntity.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiChatCompletionResponseEntity.java:81]: I actually parsed an array in my first iteration, but it felt like it doesn't fit the "one prompt and one output" interface of the API. I plan to do a follow-up for thinking about multiple inputs and what they would imply.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionRequestTaskSettings.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionRequestTaskSettings.java:43]: Adjusted with [4662370](https://github.com/elastic/elasticsearch/pull/106476/commits/466237071f03e262603bad5f27d4c7902f241eff). I've only adapted this one instance here to not clutter the PR with more changed files. I'm creating a follow-up issue for everything I'll address later.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java:47]: Makes total sense! I've added a follow-up issue to my epic, which I'll address after merge.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/openai/OpenAiChatCompletionRequestEntity.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/request/openai/OpenAiChatCompletionRequestEntity.java:50]: Yes, as the `system` and `assistant` roles are used for other purposes (setting the behavior or providing context from a previous conversation).

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiService.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/OpenAiService.java]: Fair point, adjusted with [3d6b906](https://github.com/elastic/elasticsearch/pull/106476/commits/3d6b906e311e6929b5cfb65ef41300222b20f40f).

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java]: Adjusted with [7166abf](https://github.com/elastic/elasticsearch/pull/106476/commits/7166abf82aeecc0912a3bc7c426e17a628e2fe48). Is that what you had on your mind with adding a new transport version?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiChatCompletionResponseEntity.java
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiChatCompletionResponseEntity.java:81]: > but it felt like it doesn't fit the "one prompt and one output" interface of the API. I plan to do a follow-up for thinking about multiple inputs and what they would imply.

Sounds good. I think it's fine to only expect a single item in the array result for now.

> But I guess it's safer then to set n explicitly. WDYT?

Yeah, it probably wouldn't hurt to hard code `n=1` in the request we make for now.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java
Comment (jonathan-buttner) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/openai/completion/OpenAiChatCompletionServiceSettings.java:47]: Sounds good 👍

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiChatCompletionResponseEntity.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/response/openai/OpenAiChatCompletionResponseEntity.java:81]: > Yeah, it probably wouldn't hurt to hard code n=1 in the request we make for now.

Adjusted with [a429c85](https://github.com/elastic/elasticsearch/pull/106476/commits/a429c85cfe617dfa167dcdcd6c9d327951e55052).

**Review by jonathan-buttner — COMMENTED** [review]
Thanks for those changes. I tested with a really large input text and got back an error like this:

```
"Received an unsuccessful status code for request from inference entity id [test] status [400]. Error message: [This model's maximum context length is 16385 tokens. However, your messages resulted in 16386 tokens. Please reduce the length of the messages.]"
```

For text embeddings we're looking for a slightly different message and we automatically truncate the text for now (that's an interim solution until we have chunking implemented).

How should we handle it for completions? One option, is we leave it as it is, the input text won't be truncated and the user will get a similar error back indicating they need to reduce their message. If we want that, then we should probably remove the truncation logic in the completion request (cohere doesn't leverage our internal truncation logic as an example:

```
public Request truncate() {
        return this;
    }
```

Another option would be to leverage the truncation logic we have and auto truncate the message for them. If we wanted to go that route we'll want to change this code to look for a new message: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java#L37

We could add a new boolean check here: https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java#L95 to include the other message like `contains Please reduce the length of the messages` or something like that.

**Comment by timgrein** [comment]
> How should we handle it for completions? One option, is we leave it as it is, the input text won't be truncated and the user will get a similar error back indicating they need to reduce their message.

I would say that's the way to go as it feels a little bit arbitrary to drop for example content, when you summarize something. So I would prefer to return an error and tell the user to reduce the input. I'll change that according to your suggestion.

**Comment by jimczi** [comment]
I know that @joshdevins advocates for the simplest possible API but I wonder if we should implement the [Messages API](https://huggingface.co/blog/tgi-messages-api) instead? It seems like a good standard that is applicable broadly and would ease the migration from/to this new API.

**Comment by joshdevins** [comment]
Just to clarify, I only want to avoid predefined/canned prompts. I also think that a chat/messages format API might be better as that's the direction most APIs are headed. It can be used for non-chat purposes as well so it doesn't close the door on other functionality. We discussed this briefly in [the doc](https://docs.google.com/document/d/1bttChR1ePHSF_7-imq_tLFLlixKOxr666Rmq0Muv__A/edit#heading=h.sfxzgbi0id38) but didn't come to a conclusion AFAIK.

**Comment by joshdevins** [comment]
> How should we handle it for completions?

@jonathan-buttner @timgrein regarding truncation, the APIs usually have options already for how to handle it. Shouldn't this be part of configuring the model?

**Comment by timgrein** [comment]
> Can't we offer two options? Cohere does something similar with other parameters. You can only use one of prompt or messages

To me it would make sense to offer both options. I can add `messages` in a separate PR as a follow-up as this one is already pretty big. WDYT?

**Comment by joshdevins** [comment]
If we use mutually exclusive parameters, we can always add them later. I'd continue as planned with a simple `prompt` and we add `messages` later if @jimczi and others agree to that approach. Something we can discuss in the sync tomorrow too as the ML team should have a review as the maintainers of the API.

**Comment by timgrein** [comment]
> @jonathan-buttner @timgrein regarding truncation, the APIs usually have options already for how to handle it. Shouldn't this be part of configuring the model?
There's an option [`max_tokens`](https://platform.openai.com/docs/api-reference/chat/create#chat-create-max_tokens), which limits the number of tokens in the output. The `input + output` is limited by the overall context window length, at least that's how I understood it. We can detect, whether a response was truncated by looking at the `stop_reason`. So AFAIU we would need to know the number of input tokens (by counting tokens) and output tokens (setting `max_tokens`) to know, if we need to truncate.

So if `context_window_length = 100; max_tokens = 50; input_tokens = 70` we could truncate the input by `20` on our side, which would imply that we count the tokens correctly. The tokenizers from this [library (jtokkit)](https://github.com/knuddelsgmbh/jtokkit) seem to be used by OpenAI (at least according to this [article](https://cookbook.openai.com/examples/how_to_count_tokens_with_tiktoken), which they also reference in the official docs). So we could count input tokens accurately, so that would be an option for truncating accurately (if we want to).

**Comment by joshdevins** [comment]
I'm not sure I understand the problem exactly. Why do we need to capture if the input was truncated? That's the responsibility of the user and the model provider. We should only be a proxy.

**Comment by joshdevins** [comment]
[This concerns me a lot](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java#L65-L82). We really ought not to try and decode error messages, but rather just rely on status codes. The way we inspect things now is very brittle and puts us in a position of needing to maintain compatibility across any changes the provider makes. Why do we need to do this error message inspection?

**Comment by jonathan-buttner** [comment]
> > How should we handle it for completions?
> 
> @jonathan-buttner @timgrein regarding truncation, the APIs usually have options already for how to handle it. Shouldn't this be part of configuring the model?

The truncating was mainly built as a stop gap until we have chunking when were building out the text embedding integrations. When we were adding OpenAI text embedding support, we were trying to handle the scenario where we are sending too many tokens. OpenAI will return an error indicating you need to shorten the length of the text. In that scenario we could return the error to the user and they'd need to do the chunking. What we thought might be helpful is to truncate the text and retry the request to see if we can get it small enough to retrieve the text embeddings. We also had plans to return a field in our response to indicate that it was truncated but we haven't implemented that yet 😅 

> the APIs usually have options already for how to handle it

Cohere does provide options for truncating the text using the `truncate` field. For cohere we don't truncate, we just pass the text through and return errors. My understanding is that for text embeddings for OpenAI and huggingface (for HF it probably depends on the model and I haven't done as much testing), they just return an error if you send too many tokens. And the only indication that there were too many tokens was in the error message string returned in the response.



> [This concerns me a lot](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java#L65-L82). We really ought not to try and decode error messages, but rather just rely on status codes and relaying those back to the user. The way we inspect things now is very brittle and puts us in a position of needing to maintain compatibility across any changes the provider makes. Why do we need to do this error message inspection?

Yeah fair point. I think the error message decoding is only for truncation which I think we're planning on removing once we have chunking support. Most of those if-blocks are just looking at the status codes and returning a generic error message + error message returned by the 3rd party service.

**Comment by timgrein** [comment]
> [This concerns me a lot](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java#L65-L82). We really ought not to try and decode error messages, but rather just rely on status codes and relaying those back to the user.

IMO it's worth to add this point to the sync agenda tomorrow as it's seems like a more broader topic (at least for OpenAI as a provider), which can probably be discussed separately from this PR. WDYT?

**Comment by timgrein** [comment]
> I'm not sure I understand the problem exactly. Why do we need to capture if the input was truncated? That's the responsibility of the user and the model provider. We should only be a proxy.

Our response looks like this:

```
{
    "completion": [
       "result": "inferred content..."
    ]
}
```

So currently there's no indication, if something was truncated or not. We could also simply return more information (token usage, stop reason etc.) from OpenAI's chat completion API without any adaption on our side, which would fit the "simply a proxy" model. WDYT?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java:39]: I think we can remove this now right?

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiChatCompletionResponseHandler.java
Comment (jonathan-buttner) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiChatCompletionResponseHandler.java:23]: Since I suppose it's possible that the 429 could happen if we hit too many requests per minute, maybe we just return a generic 429 message that we have: `RATE_LIMIT` is the variable name.

**Review by jonathan-buttner — APPROVED** [review]
Thanks for the changes! Looks good. Just a few nit comments.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java
Comment (timgrein) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/external/openai/OpenAiResponseHandler.java:39]: Good spot, interesting that `spotlessApply` doesn't do this for me though

**Pull request merged**
